File: src/main/java/net/hectus/neobb/turn/legacy_game/structure/LTRedstoneBlockWall.java
package net.hectus.neobb.turn.legacy_game.structure;

import net.hectus.neobb.buff.Buff;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.structure.PlacedStructure;
import net.hectus.neobb.structure.Structure;
import net.hectus.neobb.structure.StructureManager;
import net.hectus.neobb.turn.default_game.attributes.clazz.RedstoneClazz;
import net.hectus.neobb.turn.default_game.attributes.function.BuffFunction;
import net.hectus.neobb.turn.default_game.structure.StructureTurn;
import net.hectus.neobb.util.Modifiers;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.util.Vector;

import java.util.List;

public class LTRedstoneBlockWall extends StructureTurn implements BuffFunction, RedstoneClazz {
    public LTRedstoneBlockWall(NeoPlayer player) { super(player); }
    public LTRedstoneBlockWall(PlacedStructure data, NeoPlayer player) { super(data, player); }

    @Override
    public int cost() {
        return 6;
    }

    @Override
    public Structure referenceStructure() {
        return StructureManager.structure("legacy-redstone_block_wall");
    }

    @Override
    public List<ItemStack> items() {
        return List.of(new ItemStack(Material.REDSTONE_BLOCK, 14));
    }

    @Override
    public void apply() {
        Player nextPlayer = player.nextPlayer().player;
        nextPlayer.setVelocity(nextPlayer.getVelocity().clone().add(new Vector(0, 2, 0)));

        player.game.addModifier(Modifiers.G_LEGACY_NO_ATTACK);
        player.game.turnScheduler.runTaskLater("redstone_block_wall", () -> {
            player.game.removeModifier(Modifiers.G_LEGACY_NO_ATTACK);
        }, 2);
    }

    @Override
    public List<Buff> buffs() {
        return List.of(new Buff.ExtraTurn(Buff.BuffTarget.YOU, 2));
    }
}

File: src/main/kotlin/net/hectus/neobb/turn/legacy_game/structure/LTRedstoneBlockWall.kt
package net.hectus.neobb.turn.legacy_game.structure

import com.marcpg.libpg.storing.Cord
import net.hectus.neobb.buff.Buff
import net.hectus.neobb.buff.ExtraTurn
import net.hectus.neobb.game.util.ScheduleID
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.structure.PlacedStructure
import net.hectus.neobb.structure.StaticStructure
import net.hectus.neobb.structure.StaticStructures
import net.hectus.neobb.turn.default_game.attribute.clazz.NatureClazz
import net.hectus.neobb.turn.default_game.attribute.function.BuffFunction
import net.hectus.neobb.turn.default_game.structure.StructureTurn
import net.hectus.neobb.util.Modifiers
import org.bukkit.util.Vector

class LTRedstoneBlockWall(data: PlacedStructure?, cord: Cord?, player: NeoPlayer?) : StructureTurn(data, cord, player), BuffFunction, NatureClazz {
    override val cost: Int = 6
    override val staticStructure: StaticStructure = StaticStructures.Legacy.REDSTONE_BLOCK_WALL

    override fun apply() {
        val nextPlayer = player!!.nextPlayer().player
        nextPlayer.velocity = nextPlayer.velocity.clone().add(Vector(0, 2, 0))

        player.game.addModifier(Modifiers.Game.Legacy.NO_ATTACK)
        player.game.turnScheduler.runTaskLater(ScheduleID.REDSTONE_BLOCK_WALL, 2) {
            player.game.removeModifier(Modifiers.Game.Legacy.NO_ATTACK)
        }
    }

    override fun buffs(): List<Buff<*>> = listOf(ExtraTurn(2))
}
